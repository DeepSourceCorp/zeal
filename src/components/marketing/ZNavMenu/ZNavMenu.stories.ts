import ZNavMenu from './ZNavMenu.vue'

export default {
  title: 'Zeal Marketing/Nav Menu',
  component: ZNavMenu,
  excludeStories: /.*Data$/,
  parameters: {
    layout: 'fullscreen'
  }
}

export const Basic = () => ({
  components: { ZNavMenu },
  template: `<div class='bg-gradient-skeleton'>
      <z-nav-menu></z-nav-menu>
    </div>`
})
