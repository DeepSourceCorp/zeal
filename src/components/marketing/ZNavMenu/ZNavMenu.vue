<template>
  <span class="text-vanilla-100">
    <!-- Dropdown -->
    <span v-if="type === TYPES.dropdown" class="relative inline-block text-left">
      <div
        @mouseover.stop.prevent="toggleDropdown(true)"
        @mouseleave.stop.prevent="toggleDropdown(false)"
        @click.stop.prevent
        class="relative py-3 h-full"
      >
        <z-button class="px-4" type="link" :class="[`${isDropdownOpen ? 'bg-ink-300' : ''}`]">
          {{ title }}
        </z-button>
        <span
          class="absolute -ml-2 mt-1 left-1/2 block w-0 h-0 border-t-0 border-r-8 border-l-8 border-b-8 border-l-transparent border-r-transparent border-b-ink-300"
          :class="[dropdownAnimations]"
        ></span>
      </div>
      <div
        v-if="$slots.default"
        :class="[dropdownAnimations, ARRANGEMENT[arrange].classes]"
        @mouseover="toggleDropdown(true)"
        @mouseleave="toggleDropdown(false)"
        data-before=""
        class="absolute -left-12 origin-top-right bg-ink-300 rounded-md shadow-lg"
      >
        <slot></slot>
      </div>
    </span>

    <!-- Accordion -->
    <template v-if="type === TYPES.accordion">
      <span
        @click="toggleAccordion()"
        class="flex items-center cursor-pointer px-4 py-3 bg-ink-300 inline-block hover:opacity-75 select-none font-semibold w-full"
      >
        <span class="flex flex-1">{{ title }}</span>
        <z-icon icon="chevron-down" class="transform" :class="[accordionHeaderAnimations]"></z-icon>
      </span>
      <span
        class="block bg-ink-300 overflow-scroll transition-max-height duration-500 ease-in-out"
        :class="[accordionBodyAnimations]"
      >
        <span class="border-t border-ink-200 block text-sm md:text-md">
          <slot></slot>
        </span>
      </span>
    </template>

    <!-- None -->
    <span v-if="type === TYPES.none" class="block py-4">
      <span class="block font-semibold px-4">{{ title }}</span>
      <span class="block text-sm md:text-md -mb-2 -mx-1">
        <slot></slot>
      </span>
    </span>
  </span>
</template>

<script lang="ts">
import Vue from 'vue'
import ZButton from '@/components/ZButton'
import ZIcon from '@/components/ZIcon'

const TYPES = {
  accordion: 'accordion',
  dropdown: 'dropdown',
  none: 'none'
}

const ARRANGEMENT = {
  vertical: {
    text: 'vertical',
    classes: ''
  },
  horizontal: {
    text: 'horizontal',
    classes: 'flex'
  }
}

export default Vue.extend({
  components: { ZButton, ZIcon },
  name: 'ZNavMenu',
  props: {
    title: {
      type: String,
      required: true
    },
    type: {
      type: String,
      default: TYPES.none,
      validator: (type) => Object.keys(TYPES).includes(type)
    },
    arrange: {
      type: String,
      default: ARRANGEMENT.vertical.text,
      validator: (arrange) => Object.keys(ARRANGEMENT).includes(arrange)
    }
  },
  data() {
    return {
      isAccordionOpen: false,
      isDropdownOpen: false,
      accordionHeaderAnimations: '',
      TYPES,
      ARRANGEMENT
    }
  },
  computed: {
    dropdownAnimations(): string {
      return this.isDropdownOpen ? 'animate-fadeInFast' : 'hidden'
    },
    accordionBodyAnimations(): string {
      return this.isAccordionOpen ? 'max-h-102' : 'max-h-0'
    }
  },
  methods: {
    toggleDropdown(show: boolean) {
      this.isDropdownOpen = show
    },
    toggleAccordion() {
      this.isAccordionOpen = !this.isAccordionOpen
      this.accordionHeaderAnimations = this.isAccordionOpen
        ? 'animate-first-half-spin rotate-180'
        : 'animate-reverse-half-spin rotate-0'
    }
  }
})
</script>
