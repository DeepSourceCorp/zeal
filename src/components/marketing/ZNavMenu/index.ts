import ZNavMenu from './ZNavMenu.vue'

export default ZNavMenu
