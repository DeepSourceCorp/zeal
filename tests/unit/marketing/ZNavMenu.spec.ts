import Vue from 'vue'
import { createLocalVue, mount, Wrapper } from '@vue/test-utils'

import ZNavMenu from '../../../src/components/marketing/ZNavMenu'

describe('ZNavMenu component', () => {
  let mountFunction: (options?: Record<string, unknown>) => Wrapper<Vue>
  let localVue: typeof Vue
  beforeEach(() => {
    localVue = createLocalVue()
    mountFunction = (options = {}) => {
      return mount(ZNavMenu, {
        localVue,
        ...options
      })
    }
  })
  it('renders a ZNavMenu when the component is mounted', () => {
    let mountedInstance = mountFunction({
      propsData: {
        title: 'Dummy title'
      }
    })
    expect(mountedInstance.html()).toMatchSnapshot()
  })
  it('when type is dropdown', () => {
    let mountedInstance = mountFunction({
      propsData: {
        title: 'Dummy title',
        type: 'dropdown'
      },
      slots: {
        default: ['Something']
      }
    })
    expect(mountedInstance.html()).toMatchSnapshot()
    expect(mountedInstance.html()).toContain('absolute')
    expect(mountedInstance.text()).toContain('Something')
  })
  it('when type is accordion', () => {
    let mountedInstance = mountFunction({
      propsData: {
        title: 'Dummy title',
        type: 'accordion'
      },
      slots: {
        default: ['Something']
      }
    })
    expect(mountedInstance.html()).toContain('chevron-down')
    expect(mountedInstance.text()).toContain('Something')
  })
  it('when type is none', () => {
    let mountedInstance = mountFunction({
      propsData: {
        title: 'Dummy title'
      },
      slots: {
        default: ['Something']
      }
    })
    expect(mountedInstance.html()).toMatchSnapshot()
    expect(mountedInstance.html()).toContain('Dummy title')
    expect(mountedInstance.text()).toContain('Something')
  })
})
