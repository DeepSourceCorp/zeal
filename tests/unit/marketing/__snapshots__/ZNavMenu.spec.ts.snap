// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZNavMenu component renders a ZNavMenu when the component is mounted 1`] = `<span class="text-vanilla-100"><!----> <!----> <span class="block py-4"><span class="block font-semibold px-4">Dummy title</span> <span class="block text-sm md:text-md -mb-2 -mx-1"></span></span></span>`;

exports[`ZNavMenu component when type is accordion 1`] = `
<span class="text-vanilla-100"><!----> <span class="flex items-center cursor-pointer px-4 py-3 bg-ink-300 inline-block hover:opacity-75 select-none font-semibold w-full"><span class="flex flex-1">Dummy title</span> <svg viewBox="0 0 24 24" fill="none" class="stroke-2 transform z-icon w-4 h-4 sm:w-6 sm:h-6 z-icon--chevron-down text-vanilla-400">
  <polyline points="6 9 12 15 18 9"></polyline>
</svg></span> <span class="block bg-ink-300 overflow-scroll transition-max-height duration-500 ease-in-out max-h-0"><span class="border-t border-ink-200 block text-sm md:text-md">Something</span></span>
<!----></span>
`;

exports[`ZNavMenu component when type is dropdown 1`] = `
<span class="text-vanilla-100"><span class="relative inline-block text-left"><div class="relative py-3 h-full"><a href="#" type="link" class="z-btn inline-flex items-center font-medium border-transparent rounded-sm relative justify-center px-4 false false p-0 h-8 text-xs sm:h-10 sm:text-base">
        Dummy title
      </a> <span class="absolute -ml-2 mt-1 left-1/2 block w-0 h-0 border-t-0 border-r-8 border-l-8 border-b-8 border-l-transparent border-r-transparent border-b-ink-300 hidden"></span></div>
<div data-before="" class="absolute -left-12 origin-top-right bg-ink-300 rounded-md shadow-lg min-w-52 hidden">Something</div>
</span>
<!---->
<!----></span>
`;

exports[`ZNavMenu component when type is none 1`] = `<span class="text-vanilla-100"><!----> <!----> <span class="block py-4"><span class="block font-semibold px-4">Dummy title</span> <span class="block text-sm md:text-md -mb-2 -mx-1">Something</span></span></span>`;
